pkg/agent/multicast/mcast_controller.go

```diff
@@ -0,0 +1,362 @@
+			}
+		}
+	}
+	return
+}
+
+// checkLastMember sends out a query message on the group to check if there are still members in the group. If no new
+// membership report is received in the max response time, the group is removed from groupCache.
+func (c *Controller) checkLastMember(group net.IP) {
+	err := c.igmpSnooper.queryIGMP(group, queryVersion)
+	if err != nil {
+		klog.ErrorS(err, "Failed to send IGMP query message", "group", group.String())
+		return
+	}
+	c.queue.AddAfter(group.String(), igmpMaxResponseTime)
+}
+
+// clearStaleGroups removes the stale groups from groupCache. A stale group is the one that has not been updated for
+// mcastGroupTimeout.
+func (c *Controller) clearStaleGroups() {
+	now := time.Now()
+	for _, obj := range c.groupCache.List() {
+		status := obj.(*GroupMemberStatus)
+		status.mutex.RLock()
+		diff := now.Sub(status.lastProbe).Seconds()
+		status.mutex.RUnlock()
+		if diff > mcastGroupTimeout.Seconds() {
+			c.queue.Add(status.group.String())
+		}
+	}
+}
+
+type Controller struct {
+	ofClient     openflow.Client
+	nodeConfig   *config.NodeConfig
+	igmpSnooper  *IGMPSnooper
+	groupEventCh chan *mcastGroupEvent
+	groupCache   cache.Indexer
+	queue        workqueue.RateLimitingInterface
+	// installedGroups saves the groups which are configured on both OVS and the host.
+	installedGroups      sets.String
+	installedGroupsMutex sync.RWMutex
+}
+
+func NewMulticastController(ofClient openflow.Client, nodeConfig *config.NodeConfig, ifaceStore interfacestore.InterfaceStore) *Controller {
+	eventCh := make(chan *mcastGroupEvent, workerCount)
+	groupSnooper := newSnooper(ofClient, ifaceStore, eventCh)
+	groupCache := cache.NewIndexer(getGroupEventKey, cache.Indexers{
+		podInterfaceIndex: podInterfaceIndexFunc,
+	})
+	return &Controller{
+		ofClient:        ofClient,
+		nodeConfig:      nodeConfig,
+		igmpSnooper:     groupSnooper,
+		groupEventCh:    eventCh,
+		groupCache:      groupCache,
+		installedGroups: sets.NewString(),
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "multicastgroup"),
+	}
+}
+
+func (c *Controller) Initialize() error {
+	// Install flows on OVS for IGMP packets and multicast traffic forwarding:
+	// 1) send the IGMP report messages to Antrea Agent,
+	// 2) forward the IMGP query messages to all local Pods,
+	// 3) forward the multicast traffic to antrea-gw0 if no local Pods have joined in the group, and this is to ensure
+	//    local Pods can access the external multicast receivers.
+	err := c.ofClient.InstallMulticastInitialFlows(uint8(openflow.PacketInReasonMC))
+	if err != nil {
+		klog.ErrorS(err, "Failed to install multicast initial flows")
+		return err
+	}
+	return nil
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	// Periodically query Multicast Groups on OVS.
+	go wait.NonSlidingUntil(func() {
+		if err := c.igmpSnooper.queryIGMP(net.IPv4zero, queryVersion); err != nil {
```

Comment:
What's the purpose of this query? Could it get the same effect as `checkLastMember`? If yes, do we still need the latter?